<HTML>
<HEAD>
   <META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
   <META NAME="GENERATOR" CONTENT="Mozilla/4.05 [en] (X11; I; Linux 2.0.35 i686) [Netscape]">
   <TITLE>Swissknife</TITLE>
</HEAD>
<BODY>
<img align="RIGHT" src="SKlogo.gif">
<H1>
Swissknife</H1>

<H4>
An object-oriented Perl library to handle Swiss-Prot entries</H4>
Swissknife has been developed in the Swiss-Prot groups at the <a href="https://www.ebi.ac.uk">European Bioinformatics Institute</a> and the <a href="https://www.sib.swiss">Swiss Institute of Bioinformatics</a>.
<p>

The latest release is always available from <a href="https://sourceforge.net/projects/swissknife/files/latest/download">https://sourceforge.net/projects/swissknife/files/latest/download</a>.<br>
 The current and development version are hosted at 
<A href="https://sourceforge.net">
<IMG src="https://sourceforge.net/sflogo.php?group_id=31203&type=1" width="88" height="31" border="0" alt="SourceForge" align=middle></A>
[<a href="https://sourceforge.net/projects/swissknife">Swissknife project page</a>]
[<a href="https://sourceforge.net/p/swissknife/git/ci/master/tree/">Browse Git tree</a>].
<p>
General information on installation etc. is contained in the file <a href="README">README</a>. This document is the starting
point to the usage documentation of the Swissknife modules.

<h3>Usage</h3>
To use Swissknife, include the line <br><code>use SWISS::Entry;</code><br>
in your program.
<p> 
A small program using Swissknife is <a href="../examples/example.pl">example.pl</a>.
<p>
The program <a href="../examples/benchmark.pl">benchmark.pl</a> can be used to test
the Swissknife components and to give a rough idea of the
system performance. The program can be called with the Swiss-Prot
example file <a href="../examples/SWISS100.dat">SWISS100.dat</a>. <br>
Usage example:<br>
<code>cd SWISS/examples</code><br>
<code>perl benchmark.pl -file SWISS100.dat -repeats 10</code>
<p>
The output should look similar to:
<pre>*** Swissknife Benchmark and Test suite *** 
Read only:             :  0 wallclock secs ( 0.03 usr +  0.01 sys =  0.04 CPU) @ 250.00/s (n=10)
            (warning: too few iterations for a reliable count)
Read/Write NULL:       :  0 wallclock secs ( 0.03 usr +  0.01 sys =  0.04 CPU) @ 250.00/s (n=10)
            (warning: too few iterations for a reliable count)
Read/Write:            :  0 wallclock secs ( 0.03 usr +  0.03 sys =  0.06 CPU) @ 166.67/s (n=10)
            (warning: too few iterations for a reliable count)
Read/Write/addAC:      :  0 wallclock secs ( 0.11 usr +  0.03 sys =  0.14 CPU) @ 71.43/s (n=10)
            (warning: too few iterations for a reliable count)
Read/Write/Fullparse:  : 10 wallclock secs ( 9.29 usr +  0.05 sys =  9.34 CPU) @  1.07/s (n=10)
Read/Write/Fp/Update:  : 14 wallclock secs (14.22 usr +  0.04 sys = 14.26 CPU) @  0.70/s (n=10)
Read/equals:           :  0 wallclock secs ( 0.26 usr +  0.01 sys =  0.27 CPU) @ 37.04/s (n=10)
            (warning: too few iterations for a reliable count)
Read/Write/Modify:     :  4 wallclock secs ( 3.71 usr +  0.04 sys =  3.75 CPU) @  2.67/s (n=10)
</pre>
<p>
A more comprehensive test set is provided in the t/ directory:
<pre><code>cd SWISS/t
perl test.pl *.t
</code>
</pre>

should produce an output similar to
<pre><code>
*** Swissknife test suite ***

DEs.t ............. ok   
FTId.t ............ ok   
GNs.t ............. ok   
annot.t ........... ok   
crc64.t ........... ok   
evidence.t ........ ok   
fasta.t ........... ok   
formatProblems.t .. ok   
identity.t ........ ok   
util.t ............ ok   
All tests successful.
Files=10, Tests=20,  2 wallclock secs ( 0.05 usr  0.02 sys +  1.39 cusr  0.09 csys =  1.55 CPU)
Result: PASS
</code>
</pre>

<h3>Modules</h3>
<TABLE BORDER >
<TR>
<TH>Module</TH>

<TH>Documentation</TH>

<TH>Comment</TH>
</TR>

<TR>
<TH colspan=3 align=left>The main module</th>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/Entry.pm">Entry.pm</A></TD>

<TD><A HREF="docs/Entry.html">Entry.html</A></TD>

<TD>The main module to handle Swiss-Prot entries. One Entry object represents one Swiss-Prot
entry and provides an API for its modification.</TD>
</TR>

<TR>
<TH colspan=2 align=left>Line objects</th>
<td> Each line object implements a class to handle one line object of an entry or (e.g. Ref.pm) a group of related line objects.
</TR>

<TR>
<TD><A HREF="../lib/SWISS/ACs.pm">ACs.pm</A></TD>

<TD><A HREF="docs/ACs.html">ACs.html</A></TD>

<TD>Representation of the AC line.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/DTs.pm">DTs.pm</A></TD>

<TD><A HREF="docs/DTs.html">DTs.html</A></TD>

<TD>The date lines.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/DEs.pm">DEs.pm</A></TD>

<TD><A HREF="docs/DEs.html">DEs.html</A></TD>

<TD>The description lines.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/DE.pm">DE.pm</A></TD>

<TD><A HREF="docs/DE.html">DE.html</A></TD>

<TD>A single name for the protein.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/DRs.pm">DRs.pm</A></TD>

<TD><A HREF="docs/DRs.html">DRs.html</A></TD>

<TD>The DR lines, crossreferences to other databases.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/CCs.pm">CCs.pm</A>&nbsp;</TD>

<TD><A HREF="docs/CCs.html">CCs.html</A>&nbsp;</TD>


<TD>Comment lines.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/CCcopyright.pm">CCcopyright.pm</A></TD>

<TD><A HREF="docs/CCcopyright.html">CCcopyright.html</A></TD>

<TD>The copyright statement (part of the comment lines).</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/CCalt_prod.pm">CCalt_prod.pm</A></TD>

<TD><A HREF="docs/CCalt_prod.html">CCalt_prod.html</A></TD>

<TD>One comment object of the topic ALTERNATIVE PRODUCTS.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/CCrna_editing.pm">CCrna_editing.pm</A></TD>

<TD><A HREF="docs/CCrna_editing.html">CCrna_editing.html</A></TD>

<TD>One comment object of the topic RNA EDITING.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/CCbpc_properties.pm">CCbpc_properties.pm</A></TD>

<TD><A HREF="docs/CCbpc_properties.html">CCbpc_properties.html</A></TD>

<TD>One comment object of the topic BIOPHYSICOCHEMICAL PROPERTIES.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/CCinteraction.pm">CCinteraction.pm</A></TD>

<TD><A HREF="docs/CCinteraction.html">CCinteraction.html</A></TD>

<TD>One comment object of the topic INTERACTION.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/CCdisease.pm">CCdisease.pm</A></TD>

<TD><A HREF="docs/CCdisease.html">CCdisease.html</A></TD>

<TD>One comment object of the topic DISEASE.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/CCsubcell_location.pm">CCsubcell_location.pm</A></TD>

<TD><A HREF="docs/CCsubcell_location.html">CCsubcell_location.html</A></TD>

<TD>One comment object of the topic SUBCELLULAR LOCATION.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/CC.pm">CC.pm</A></TD>

<TD><A HREF="docs/CC.html">CC.html</A></TD>

<TD>One comment object of any other topic.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/FTs.pm">FTs.pm</A></TD>

<TD><A HREF="docs/FTs.html">FTs.html</A></TD>

<TD>The feature lines.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/GNs.pm">GNs.pm</A></TD>

<TD><A HREF="docs/GNs.html">GNs.html</A></TD>

<TD>The gene lines.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/GeneGroup.pm">GeneGroup.pm</A></TD>

<TD><A HREF="docs/GeneGroup.html">GeneGroup.html</A></TD>

<TD>The different synonyms for a single gene name.</TD>
</TR>
<TR>
<TD><A HREF="../lib/SWISS/GN.pm">GN.pm</A></TD>

<TD><A HREF="docs/GN.html">GN.html</A></TD>

<TD>One single gene name.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/IDs.pm">IDs.pm</A></TD>

<TD><a href="docs/IDs.html">IDs.html</a></TD>

<TD>The ID line.</TD>
</TR>
 
<TR>
<TD><A HREF="../lib/SWISS/KWs.pm">KWs.pm</A></TD>

<TD><A HREF="docs/KWs.html">KWs.html</A></TD>

<TD>The keyword lines. KWs is a container object, it holds an array of
KW objects.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/KW.pm">KW.pm</A></TD>

<TD><A HREF="docs/KW.html">KW.html</A></TD>

<TD>One keyword object.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/OCs.pm">OCs.pm</A></TD>

<TD><a href="docs/OCs.html">OCs.html</a></TD>

<TD>The OC line encoding the taxonomy of the source organism.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/OGs.pm">OGs.pm</A></TD>

<TD><A HREF="docs/OGs.html">OGs.html</A></TD>

<TD>The OG lines. OGs is a container object, it holds an array of
OG objects.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/OG.pm">OG.pm</A></TD>

<TD><A HREF="docs/OG.html">OG.html</A></TD>

<TD>One organism name.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/OSs.pm">OSs.pm</A></TD>

<TD><A HREF="docs/OSs.html">OSs.html</A></TD>

<TD>The OS lines. OSs is a container object, it holds an array of
OS objects.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/OS.pm">OS.pm</A></TD>

<TD><A HREF="docs/OS.html">OS.html</A></TD>

<TD>One organism name.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/OXs.pm">OXs.pm</A></TD>

<TD><A HREF="docs/OXs.html">OXs.html</A></TD>

<TD>The OX lines. OXs is a container object, for each valid taxonomic resource it contains a ListBase object which holds a list of OX objects.
</TR>

<TR>
<TD><A HREF="../lib/SWISS/OX.pm">OX.pm</A></TD>

<TD><A HREF="docs/OX.html">OX.html</A></TD>

<TD>One tax id object.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/Ref.pm">Ref.pm</A></TD>

<TD><A HREF="docs/Ref.html">Ref.html</A></TD>

<TD>Represents one literature reference.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/Refs.pm">Refs.pm</A></TD>

<TD><A HREF="docs/Refs.html">Refs.html</A></TD>

<TD>Represents the list of literature references.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/Stars.pm">Stars.pm</A></TD>

<TD><A HREF="docs/Stars.html">Stars.html</a></TD>

<TD>The "annotator's section" (See <a href="#stars">note</a>)</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/Stars/aa.pm">Stars/aa.pm</A></TD>

<TD><A HREF="docs/Stars/aa.html">Stars/aa.html</A></TD>

<TD>Unstructured notes in the internal "annotator's section". See <A HREF="Stars.html">Stars.html</a></TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/Stars/DR.pm">Stars/DR.pm</A></TD>

<TD><A HREF="docs/Stars/DR.html">Stars/DR.html</A></TD>

<TD>DR in the internal "annotator's section". See <A HREF="Stars.html">Stars.html</a></TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/Stars/EV.pm">Stars/EV.pm</A></TD>

<TD><A HREF="docs/Stars/EV.html">Stars/EV.html</A></TD>

<TD>Evidence section. See <A HREF="Stars.html">Stars.html</a></TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/Stars/default.pm">Stars/default.pm</A></TD>

<TD><A HREF="docs/Stars/default.html">Stars/default.html</A></TD>

<TD>Default class for structured information in the internal "annotator's section". See <A HREF="Stars.html">Stars.html</a></TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/SQs.pm">SQs.pm</A></TD>

<TD><A HREF="docs/SQs.html">SQs.html</A></TD>

<TD>The sequence lines.</TD>
</TR>

<TR>
<TH colspan=2 align=left>Base objects</th>
<td>These modules implement the base classes from which all line object classes are derived.
</TR>

<TR>
<TD><A HREF="../lib/SWISS/BaseClass.pm">BaseClass.pm</A></TD>

<TD><A HREF="docs/BaseClass.html">BaseClass.html</A></TD>

<TD>The base class, implementing common methods, e.g. equals</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/ListBase.pm">ListBase.pm</A></TD>

<TD><A HREF="docs/ListBase.html">ListBase.html</A></TD>

<TD>Provides methods to manipulate list-based objects like KWs.pm</TD>
</TR>

<TR>
<TH colspan=2 align=left>Auxiliary modules</th>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/TextFunc.pm">TextFunc.pm</A></TD>

<TD><A HREF="docs/TextFunc.html">TextFunc.html</A></TD>

<TD>Auxiliary functions, mainly for text formatting.</TD>
</TR>

<TR>
<TD><A HREF="../lib/SWISS/CRC64.pm">CRC64.pm</A></TD>

<TD><A HREF="docs/CRC64.html">CRC64.html</A></TD>

<TD>Provides a method to calculate the CRC64 checksum.
</TD>
</TR>

</TABLE>

<h3>Bugs, Feedback</h3>
The Swissknife modules have been developed for internal use and are
provided as they are. However, if they are actually used by
external users, we'll happily try to incoporate any suggestions
for improvement (especially on the documentation side?). Therefore: 
<p>
<em>Please report any bugs and suggestions for improvement</em> to sk <i>at</i> ebi <i>dot</i> ac <i>dot</i> uk.


<h3>Notes</h3>
<a name="stars"></a>
<h4>The ** lines (Stars.pm)</h4>
The Swissknife modules are used in the production of the TrEMBL
protein database in the Swiss-Prot group at the EBI. The internal
version of the entries may contain additional information for the
database curators. This information is stored in lines with the line
tag '**'. Therefore the Swissknife modules provide methods to handle
these lines, although they are not visible to the public. As the ** lines
may also be used to store additional information of the external
users, the corresponding methods are not removed for the public
release.

<a name="evidenceTags"><h4>Evidence tags</h4>
From June 2000 onwards, we are introducing <a
href="http://www3.ebi.ac.uk/~sp/intern/projects/evidenceTags/">evidence
tags</a> into UniProtKB/Trembl. In the beginning, these will be deleted from the public version. However, Swissknife provides functions to handle
them. See <A HREF="docs/ListBase.html">ListBase.html</A> and <A
HREF="docs/BaseClass.html">BaseClass.html</A>. Making Swissknife
"Evidence tag compatible" also required a major change to the
interface of the <A HREF="docs/KWs.html">KWs</A>, <A HREF="docs/OGs.html">OGs</A> and <A
HREF="docs/OSs.html">OSs</A> modules. Originally they were simple
ListBase classes, where each keyword/organism name was one element of
the ListBase array. Now each keyword/organism name is held in its own
object, see <A HREF="docs/KW.html">KW.html</A>, <A HREF="docs/OG.html">OG.html</A> and <A HREF="docs/OS.html">OS.html</A>. <a href="examples/evTest.pl">evTest.pl</a> is a sample program manipulating evidence tags.

<h4>Evidences in the UniProtKB flat file format</h4>
(Not public in UniProtKB before October 1, 2014)
The evidence for annotations in UniProtKB entries has been available for several years in the XML and RDF representation of the data and we now intend to add this information to the text format (aka flat file format).
Swissknife handles both those new evidences (in form {ECO:...[, ECO:...]}) and the old ones...


<h3>Authors</h3>
<i>Swissknife has been developed by:</i><br>
Wolfgang Fleischmann (European Bioinformatics Institute)<br>
Alexandre Gattiker (Swiss Institute of Bioinformatics)<br>
Henning Hermjakob (European Bioinformatics Institute)<br>
Eric Jain (Swiss Institute of Bioinformatics)<br>
Paul Kersey (European Bioinformatics Institute)<br>
Edouard de Castro (Swiss Institute of Bioinformatics)<br>
<br>
</ADDRESS>
</BODY>
</HTML>
